File: LeetCode/0085-maximal-rectangle/0085-maximal-rectangle.java
class Solution {
    public int[] getNSR(int[] height) {
        Stack<Integer> st = new Stack<>();
        int n = height.length;
        int[] NSR = new int[n];
        for (int i = n - 1; i >= 0; i--) {
            if (st.empty()) {
                NSR[i] = n;
            } else {
                while (!st.empty() && height[st.peek()] >= height[i]) {
                    st.pop();
                }
                if (st.empty()) {
                    NSR[i] = n;
                } else {
                    NSR[i] = st.peek();
                }
            }
            st.push(i);
        }
        return NSR;
    }

    public int[] getNSL(int[] height) {
        Stack<Integer> st = new Stack<>();
        int n = height.length;
        int[] NSL = new int[n];
        for (int i = 0; i < n; i++) {
            if (st.empty()) {
                NSL[i] = -1;
            } else {
                while (!st.empty() && height[st.peek()] >= height[i]) {
                    st.pop();
                }
                if (st.empty()) {
                    NSL[i] = -1;
                } else {
                    NSL[i] = st.peek();
                }
            }
            st.push(i);
        }
        return NSL;
    }

    public int findMaxArea(int[] height) {
        int[] NSR = getNSR(height);
        int[] NSL = getNSL(height);
        int n = height.length;
        int[] width = new int[n];
        for (int i = 0; i < n; i++) {
            width[i] = NSR[i] - NSL[i] - 1;
        }
        int maxArea = 0;
        for (int i = 0; i < n; i++) {
            int a = width[i] * height[i];
            maxArea = Math.max(maxArea, a);
        }
        return maxArea;
    }

    public int maximalRectangle(char[][] matrix) {
        if (matrix.length == 0) {
            return 0;
        }
        int m = matrix[0].length, n = matrix.length;
        int[] height = new int[m];
        for (int i = 0; i < m; i++) {
            height[i] = (matrix[0][i] == '1') ? 1 : 0;
        }
        int maxArea = findMaxArea(height);
        for (int row = 1; row < n; row++) {
            for (int col = 0; col < m; col++) {
                if (matrix[row][col] == '0') {
                    height[col] = 0;
                } else {
                    height[col] += 1;
                }
            }
            maxArea = Math.max(maxArea, findMaxArea(height));
        }
        return maxArea;
    }
}

File: LeetCode/0085-maximal-rectangle/0085-maximal-rectangle.py
class Solution:
    def getNSR(self, height: List[int]) -> List[int]:
        stack = deque()
        n = len(height)
        NSR = [0] * n
        for i in range(n - 1, -1, -1):
            while stack and height[stack[-1]] >= height[i]:
                stack.pop()
            NSR[i] = stack[-1] if stack else n
            stack.append(i)
        return NSR

    def getNSL(self, height: List[int]) -> List[int]:
        stack = deque()
        n = len(height)
        NSL = [-1] * n
        for i in range(n):
            while stack and height[stack[-1]] >= height[i]:
                stack.pop()
            NSL[i] = stack[-1] if stack else -1
            stack.append(i)
        return NSL

    def findMaxArea(self, height: List[int]) -> int:
        NSR = self.getNSR(height)
        NSL = self.getNSL(height)
        n = len(height)
        width = [0] * n
        for i in range(n):
            width[i] = NSR[i] - NSL[i] - 1
        maxArea = 0
        for i in range(n):
            a = width[i] * height[i]
            maxArea = max(maxArea, a)
        return maxArea

    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        m, n = len(matrix[0]), len(matrix)
        height = [0] * m
        maxArea = self.findMaxArea(height)
        for row in range(n):
            for col in range(m):
                height[col] = 0 if matrix[row][col] == '0' else height[col] + 1
            maxArea = max(maxArea, self.findMaxArea(height))
        return maxArea